src/rebuild/srv.c

```diff
@@ -1591,26 +1604,42 @@ ds_rebuild_schedule(struct ds_pool *pool, uint32_t map_ver,
 		return rc == 1 ? 0 : rc;
 
 	/* No existing task was found - allocate a new one and use it */
-	D_ALLOC_PTR(task);
-	if (task == NULL)
+	D_ALLOC_PTR(new_task);
+	if (new_task == NULL)
 		return -DER_NOMEM;
 
 	rc = daos_gettime_coarse(&cur_ts);
 	D_ASSERT(rc == 0);
 
-	task->dst_schedule_time = cur_ts + delay_sec;
-	task->dst_map_ver = map_ver;
-	task->dst_rebuild_op = rebuild_op;
-	uuid_copy(task->dst_pool_uuid, pool->sp_uuid);
-	D_INIT_LIST_HEAD(&task->dst_list);
+	new_task->dst_schedule_time = cur_ts + delay_sec;
+	new_task->dst_map_ver = map_ver;
+	new_task->dst_rebuild_op = rebuild_op;
+	uuid_copy(new_task->dst_pool_uuid, pool->sp_uuid);
+	D_INIT_LIST_HEAD(&new_task->dst_list);
 
 	/* TODO: Merge everything for reclaim */
-	rc = pool_target_id_list_merge(&task->dst_tgts, tgts);
+	rc = pool_target_id_list_merge(&new_task->dst_tgts, tgts);
 	if (rc)
 		D_GOTO(free, rc);
 
 	rebuild_print_list_update(pool->sp_uuid, map_ver, rebuild_op, tgts);
-	d_list_add_tail(&task->dst_list, &rebuild_gst.rg_queue_list);
+
+	/* Insert the task into the queue by order to make sure the rebuild
+	 * task with smaller version are being executed first.
+	 */
+	inserted_pos = &rebuild_gst.rg_queue_list;
+	d_list_for_each_entry(task, &rebuild_gst.rg_queue_list, dst_list) {
+		if (uuid_compare(task->dst_pool_uuid,
+				 new_task->dst_pool_uuid) != 0)
+			continue;
+
+		if (new_task->dst_map_ver > task->dst_map_ver)
+			continue;
+
+		inserted_pos = &task->dst_list;
+		break;
+	}
+	d_list_add_tail(&new_task->dst_list, inserted_pos);
```

Comment:
Should this be d_list_add?

Suppose you have 1 -> 2 -> 4 in the list and you are inserting 3.   Wouldn't adding it as tail make 1-> 2-> 4-> 3 ?

Though it is a bit inconsistent with what happens if nothing matches...then you end up pointing at the last entry

Reply:
oh, inserted_pos point to 4  in this case, then d_list_add_tail() will insert the new entry before 4.  (between 2 and 4)

static inline void
d_list_add_tail(d_list_t *newe, d_list_t *head)
{
        __gurt_list_add(newe, head->prev, head);
}

so the list will become 1->2->3->4.